java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/PriorityQueueRun.java

```diff
@@ -0,0 +1,199 @@
+                priority2Tasks.computeIfAbsent(priority, __ -> new ArrayList<>())
+                              .add(new TaskDescription(task, data, result));
+
+                scheduleNext();
+            }
+        }, delay);
+
+        return result;
+    }
+
+    private synchronized void scheduleNext() {
+        Priority foundPriority = null;
+        List<TaskDescription<?, ?>> foundTasks = null;
+
+        for (Entry<Priority, List<TaskDescription<?, ?>>> e : priority2Tasks.entrySet()) {
+            if (!e.getValue().isEmpty()) {
+                foundPriority = e.getKey();
+                foundTasks = e.getValue();
+                break;
+            }
+        }
+
+        if (foundPriority == null) {
+            return ;
+        }
+
+        if (currentPriority == null) {
+            Priority thisPriority = currentPriority = foundPriority;
+            TaskDescription thisTask = foundTasks.remove(0);
+            CancelCheck thisTaskCheck = currentTaskCheck = new CancelCheck();
+
+            WORKER.post(() -> {
+                thisTask.result.whenComplete((__, exc) -> {
+                    if (exc instanceof CancellationException) {
+                        thisTaskCheck.cancel();
+                    }
+                });
+
+                if (thisTask.result.isCancelled()) {
+                    //nothing to do anymore:
+                    return ;
+                }
+
+                Object result = null;
+                Throwable exception = null;
+
+                try {
+                    result = thisTask.task.compute(thisTask.data, thisTaskCheck);
+                } catch (Throwable t) {
+                    exception = t;
+                }
+
+                synchronized (PriorityQueueRun.this) {
+                    currentPriority = null;
+                    currentTaskCheck = null;
+
+                    if (!thisTask.result.isCancelled()) {
+                        if (exception != null) {
+                            thisTask.result.completeExceptionally(exception);
+                        } else if (thisTaskCheck.isCancelled()) {
+                            priority2Tasks.computeIfAbsent(thisPriority, __ -> new ArrayList<>())
+                                          .add(0, thisTask);
+                        } else {
+                            thisTask.result.complete(result);
+                        }
+                    }
+                }
+
+                scheduleNext();
+            });
+        }
+
+        if (currentPriority != null && currentPriority.compareTo(foundPriority) < 0) {
+            //cancel the currently running task:
+            currentTaskCheck.cancel();
+        }
+    }
+
+    private static final class TaskDescription<P, R> {
+        private final CancellableTask<P, R> task;
+        private final P data;
+        private final CompletableFuture<R> result;
+
+        public TaskDescription(CancellableTask<P, R> task, P data, CompletableFuture<R> result) {
```

Comment:
Perhaps add null checks on parameters so that NPEs happen immediately and with a useful stack trace, rather than with a delay.